### PR TITLE
[ADD] store engine: a state management system for components

### PR DESCRIPTION
**Pull request by LucasLefevre**
## Description:

This commit introduces a new way to manage state and business logic
of components based on owl reactivity.

Read `src/store_engine/README.md` for a full explanation and details.

The main goals are:

- Centralize the state used by several components across the UI.
- Move as much business logic as possible out of components (easier/faster
  to test).
- Interact with UI from the outside world (example: we currently cannot open
  a sidepanel from the outside of the app.)

Why not plugins?
===============
While these goals could be addressed by relocating state and business logic
to plugins, several disadvantages are associated with this approach:

- Excessive (deep) rendering occurs at every command dispatch, even if only a
  single component has actually changed.

- Plugins need boilerplate for getters and commands. It's also usually not
  worth the noise in getters and commands, all for the sake of a single component.

- When the state is specific to a single component, commands are only handled
  by one plugin. This means command multi-casting and shared getters doesn't
  bring any benefit.

- Unlike plugins, a single component can be mounted multiple times. Each
  component could possess its own state.
Task: : [TASK_ID](https://www.odoo.com/web#id=TASK_ID&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_t("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

**Review by rrahir — COMMENTED** [review]
comments of the day

File: src/store_engine/store.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+ * Remove the first element of a tuple
+ * @example
+ * type A = SkipFirst<[number, string, boolean]> // [string, boolean]
+ */
+type SkipFirst<T extends any[]> = T extends [any, ...infer U] ? U : never;
+
+/**
+ * TODO
+ */
+export function createMetaStore<T extends unknown>(storeName: string): StoreConstructor<T> {
```
[the remark this answers is not in the document]
Comment (LucasLefevre) [reply src/store_engine/store.ts:44]: not sure... "abstract" usually means something different (like an abstract class that you need to implement)
But I'm totally open to suggestions, "meta" isn't really great either.
`createInjectableExternalStore` ?

File: src/model.ts [hunk 2]
```diff
@@ -590,10 +593,12 @@ export class Model extends EventBus<any> implements CommandDispatcher {
    * context. This is probably the way we should do if we want to be able to
    * freeze a part of the grid (so, we would need to render different zones)
    */
-  drawGrid(context: GridRenderingContext) {
-    // we make sure here that the viewport is properly positioned: the offsets
-    // correspond exactly to a cell
-    for (let [renderer, layer] of this.renderers) {
+  drawGrid(context: GridRenderingContext, layer: LAYERS) {
```
[the remark this answers is not in the document]
Comment (LucasLefevre) [reply src/model.ts:597]: rename in all plugins

File: src/components/helpers/highlight_hook.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+  useHighlights({
+    get highlights() {
+      return hoverState.hovered ? highlightProvider.highlights : [];
+    },
+  });
+  useEffect(
+    () => {
+      env.model.dispatch("RENDER_CANVAS");
+    },
```
Comment (VincentSchippefilt) [inline src/components/helpers/highlight_hook.ts:19]: in this case, I wonder why we still need a command that will be handled by only one component and have no side effects (history). Should we not have a simple function call to model.drawGrid() or useGridDrawing ?

Comment (LucasLefevre) [reply src/components/helpers/highlight_hook.ts:19]: you're right, we don't need this command. It's only dispatched to trigger a rendering. Which can be achieved with a simple function call :+1:

Comment (LucasLefevre) [reply src/components/helpers/highlight_hook.ts:19]: mouais, not that straight forward because it doesn't have a direct access to the canvas

**Review by VincentSchippefilt — APPROVED** [review]

**Comment by VincentSchippefilt** [comment]
robodoo r+

**Comment by LucasLefevre** [comment]
robodoo rebase-ff

**Pull request closed without merging**